File: #147_Insertion_Sort_List/Solution.java
class ListNode {
    int val;
    ListNode next;
    ListNode(int x) { val = x; }
}

public class Solution {
    public ListNode insertionSortList(ListNode head) {
		if (head == null) return head;
		ListNode dummy = new ListNode(Integer.MIN_VALUE);
		dummy.next = head;
		ListNode sorted = head;
		while (sorted.next != null) { // handle only one element
			ListNode cur = sorted.next;
			/* new value is max */
			if (cur.val >= sorted.val) {
				sorted = cur;
				continue;
			}
			ListNode search = dummy.next;
			ListNode prev = dummy;
			sorted.next = cur.next;
			/* min value */
			if (cur.val <= search.val) {
				cur.next = search;
				dummy.next = cur;
				continue;
			}
			/* go thru the sorted part */
			while (search.val < cur.val) {
				prev = prev.next;
				search = search.next;
			}
			prev.next = cur;
			cur.next = search;
		}
		return dummy.next;
    }
}

File: #147_Insertion_Sort_List/Solution.py
class ListNode:
	def __init__(self, x):
		self.val = x
		self.next = None

class Solution:
	# @param {ListNode} head
	# @return {ListNode}
	def insertionSortList(self, head):
		minVal = -2147483648
		sortedHead = ListNode(minVal)
		cur = head
		maxNode = None
		while cur:
			if maxNode and maxNode.val <= cur.val:
				maxNode.next = cur
				maxNode = maxNode.next
				cur = cur.next
				maxNode.next = None
			else:
				if not maxNode:
					maxNode = cur
				prev = sortedHead
				search = prev.next
				while search and search.val < cur.val:
					search = search.next
					prev = prev.next
				tmp = cur.next
				prev.next = cur
				cur.next = search
				cur = tmp
		return sortedHead.next

	def traverse(self, head):
		while head:
			print head.val,
			head = head.next

sol = Solution()
n0 = ListNode(3)
n1 = ListNode(4)
n2 = ListNode(1)
n0.next = n1
n1.next = n2
# sol.traverse(n0)
# print
sol.traverse(sol.insertionSortList(n0))
# sol.insertionSortList(n0)